### PR TITLE
feat(ssh-keys): delete ssh keys API

### PR DESCRIPTION
**Pull request by maghirardelli**
Issue #, if available: GALI-2245

Description of changes:
backend logic, unit tests, new DELETE route, negative integ tests (more to come)

TO USE NEW SSH KEY APIS YOU NEED TO UPDATE YOUR HOSTING ACCOUNT USING THE POST /awsAccountTemplateUrls API AND CLICKING THE UPDATE LINK TO UPLOAD NEW HOSTING ACCOUNT TEMPLATE WITH PROPER EC2 KEY PAIRS PERMISSIONS.

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: solutions/swb-reference/integration-tests/support/resources/sshKeys/sshKeys.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import SshKey from './sshKey';
+
+export default class SshKeys extends CollectionResource {
+  public constructor(clientSession: ClientSession, parentApi: string = '') {
+    super(clientSession, 'sshKeys', 'sshKey');
+    const parentRoute = parentApi ? `${parentApi}/` : '';
+    this._api = `${parentRoute}sshKeys`;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (aws-tyler) [inline solutions/swb-reference/integration-tests/support/resources/sshKeys/sshKeys.ts:16]: ?

File: solutions/swb-reference/integration-tests/tests/isolated/sshKeys/delete.test.ts [hunk 2]
```diff
@@ -11,29 +12,104 @@ import { checkHttpError } from '../../../support/utils/utilities';
...
+  beforeAll(async () => {
     adminSession = await setup.getDefaultAdminSession();
+    const { data: costCenter } = await adminSession.resources.costCenters.create({
+      name: 'test cost center',
+      accountId: setup.getSettings().get('defaultHostingAccountId'),
+      description: 'a test object'
+    });
+
+    const { data } = await adminSession.resources.projects.create({
+      name: `TestProject-${uuidv4()}`,
```
Comment (aws-tyler) [inline solutions/swb-reference/integration-tests/tests/isolated/sshKeys/delete.test.ts:28]: use random name generator

File: solutions/swb-reference/src/services/sshKeyService.ts [hunk 3]
```diff
@@ -43,8 +51,63 @@ export default class SshKeyService implements SshKeyPlugin {
...
+      'Delete',
+      externalId,
+      this._aws
+    );
+
+    // get ssh key
+    let keys = [];
+    try {
+      const response = await ec2.describeKeyPairs({ Filters: [{ Name: 'key-name', Values: [sshKeyId] }] });
+      console.log(response);
```
Comment (aws-tyler) [inline solutions/swb-reference/src/services/sshKeyService.ts:71]: 👋

File: solutions/swb-reference/src/services/sshKeyService.ts [hunk 4]
```diff
@@ -91,10 +154,30 @@ export default class SshKeyService implements SshKeyPlugin {
...
-    const ec2 = hostSdk.clients.ec2;
-    const ec2InstanceConnect = hostSdk.clients.ec2InstanceConnect;
+    try {
+      const hostSdk = await aws.getAwsServiceForRole(params);
+      const ec2 = hostSdk.clients.ec2;
+      const ec2InstanceConnect = hostSdk.clients.ec2InstanceConnect;
+
+      return { ec2, ec2InstanceConnect };
+    } catch (e) {
+      throw new Error(`Could not get host EC2 clients using ${envMgmtRoleArn} and ${externalId}`);
```
Comment (aws-tyler) [inline solutions/swb-reference/src/services/sshKeyService.ts:164]: `AwsServiceError`

**Review by aws-tyler — APPROVED** [review]

File: solutions/swb-app/src/sshKeyRoutes.ts [hunk 5]
```diff
@@ -3,7 +3,52 @@
...
+        currentUserId: res.locals.user.id,
+        sshKeyId: req.params.sshKeyId,
+        projectId: req.params.projectId
+      });
+
+      try {
+        await sshKeyService.deleteSshKey(validatedResult);
+        res.status(204).send();
+      } catch (e) {
+        console.log(e);
```
[the remark this answers is not in the document]
Comment (maghirardelli) [reply solutions/swb-app/src/sshKeyRoutes.ts:33]: Fixed!

File: solutions/swb-reference/integration-tests/support/resources/sshKeys/sshKeys.ts (hunk 1 above)
Comment (maghirardelli) [reply solutions/swb-reference/integration-tests/support/resources/sshKeys/sshKeys.ts:16]: took out!

File: solutions/swb-reference/integration-tests/tests/isolated/sshKeys/delete.test.ts (hunk 2 above)
Comment (maghirardelli) [reply solutions/swb-reference/integration-tests/tests/isolated/sshKeys/delete.test.ts:28]: got it!

File: solutions/swb-reference/src/services/sshKeyService.ts (hunk 3 above)
Comment (maghirardelli) [reply solutions/swb-reference/src/services/sshKeyService.ts:71]: deleted

File: solutions/swb-reference/src/services/sshKeyService.ts (hunk 4 above)
Comment (maghirardelli) [reply solutions/swb-reference/src/services/sshKeyService.ts:164]: changed

File: solutions/swb-reference/src/services/sshKeyService.test.ts [hunk 6]
```diff
@@ -85,29 +91,160 @@ describe('SshKeyService', () => {
...
+              });
+            });
+
+            describe('and get EC2 call succeeds', () => {
+              beforeEach(() => {
+                hostEc2.deleteKeyPair = jest.fn(() => Promise.resolve({ $metadata: {} }));
+              });
+              test('it succeeds, nothing is returned', async () => {
+                // OPERATE n CHECK
+                await expect(sshKeyService.deleteSshKey(deleteSshKeyRequest)).resolves.not.toThrow();
```
Comment (janeyuaws) [inline solutions/swb-reference/src/services/sshKeyService.test.ts:242]: Do you want to check the deleteSshKey method actually delete the key (call describeKeyPairs again), or are you planning to done that in integration test

Comment (maghirardelli) [reply solutions/swb-reference/src/services/sshKeyService.test.ts:242]: This is a mocked call to `describeKeyPairs` and `deleteKeyPair`. The check you are referring to will be a multi-step integ test!

Comment (janeyuaws) [reply solutions/swb-reference/src/services/sshKeyService.test.ts:242]: got it!

**Review by aws-tyler — APPROVED** [review]

**Review by janeyuaws — APPROVED** [review]

**Pull request merged**